File: app/src/main/java/zone/com/zanimate/_ReboundActivity.java
package zone.com.zanimate;

import android.animation.ArgbEvaluator;
import android.graphics.Color;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;
import com.facebook.rebound.SimpleSpringListener;
import com.facebook.rebound.Spring;
import com.facebook.rebound.SpringChain;
import com.facebook.rebound.SpringConfig;
import com.facebook.rebound.SpringSystem;
import com.facebook.rebound.SpringUtil;
import com.zone.lib.utils.view.ViewUtils;
import com.zone.view.SquareImageView2;
import java.util.List;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class _ReboundActivity extends AppCompatActivity {

    @BindView(R.id.siv)
    SquareImageView2 siv;
    @BindView(R.id.ll_main)
    LinearLayout ll_main;
    @BindView(R.id.seekBarFriction)
    SeekBar seekBarFriction;
    @BindView(R.id.seekTension)
    SeekBar seekTension;
    @BindView(R.id.tv_Friction)
    TextView tv_Friction;
    @BindView(R.id.tv_Tension)
    TextView tv_Tension;
    private Spring spring;
    private int tensionValue = (int) SpringConfig.defaultConfig.tension;
    private int frictionValue = (int) SpringConfig.defaultConfig.friction;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.a_rebound);
        ButterKnife.bind(this);
        spring =  SpringSystem.create().createSpring();

        addReboundListener();
        updateConfig();
        updateSeekbar();

    }

    private void addReboundListener() {
        spring.addListener(new SimpleSpringListener() {
            @Override
            public void onSpringUpdate(Spring spring) {
                super.onSpringUpdate(spring);
                double currentValue = spring.getCurrentValue();
                double valueMap = SpringUtil.mapValueFromRangeToRange(currentValue, 0, 1, 1, 0.5);
                siv.setScaleX((float) valueMap);
                siv.setScaleY((float) valueMap);
            }
        });

        siv.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        spring.setEndValue(1);
                        break;
                    case MotionEvent.ACTION_UP:
                        spring.setEndValue(0);
                        break;
                }
                return true;
            }
        });
    }


    int startColor = Color.argb(255, 255, 64, 230);
    int endColor = Color.argb(255, 0, 174, 255);

    @OnClick({R.id.playTogether, R.id.play})
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.play:
                siv.bringToFront();
                spring.setEndValue(spring.getEndValue() == 0 ? 1 : 0);
                break;
            case R.id.playTogether:
                playTogether();
                break;

        }
    }

    private void playTogether() {
        ll_main.bringToFront();
        ll_main.removeAllViews();
        final SpringChain mSpringChain = SpringChain.create();
        final int offset = (ll_main.getRootView().getWidth() + ll_main.getWidth()) / 2;
        for (int i = 0; i < 7; i++) {
            final View v = new View(this);
            ArgbEvaluator argbEvaluator = new ArgbEvaluator();
            v.setBackgroundColor((Integer) argbEvaluator.evaluate((float) i / (float) 7, startColor, endColor));
            ll_main.addView(v, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1));
            v.setTranslationX(-offset);
            mSpringChain.addSpring(new SimpleSpringListener() {
                @Override
                public void onSpringUpdate(Spring spring) {
                    float value = (float) spring.getCurrentValue();
                    v.setTranslationX(value);
                }
                @Override
                public void onSpringAtRest(Spring spring) {
                    super.onSpringAtRest(spring);
                    v.setVisibility(View.INVISIBLE);
                }
            });
            if (i == 6)
                v.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        List<Spring> springList = mSpringChain.getAllSprings();
                        for (Spring spring1 : springList)
                            spring1.setCurrentValue(-offset);
                        for (int i1 = 0; i1 < ll_main.getChildCount(); i1++)
                            ViewUtils.recurrenceClipChildren(ll_main.getChildAt(i1), false);
                        mSpringChain.setControlSpringIndex(2).getControlSpring().setEndValue(0);
                    }
                }, 500);
        }
    }


    private void updateSeekbar() {
        seekTension.setOnSeekBarChangeListener(new SimpleOnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                super.onProgressChanged(seekBar, progress, fromUser);
                tensionValue = progress;
                updateConfig();
            }
        });
        seekBarFriction.setOnSeekBarChangeListener(new SimpleOnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                super.onProgressChanged(seekBar, progress, fromUser);
                frictionValue = progress;
                updateConfig();
            }
        });
    }

    private void updateConfig() {
        tv_Tension.setText(" 拉力(tension):" + tensionValue);
        tv_Friction.setText(" 摩擦力(friction):" + frictionValue);
        seekTension.setProgress(tensionValue);
        seekBarFriction.setProgress(frictionValue);
        spring.setSpringConfig(new SpringConfig(tensionValue, frictionValue));
    }


    public static class SimpleOnSeekBarChangeListener implements SeekBar.OnSeekBarChangeListener {
        @Override
        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {

        }

        @Override
        public void onStartTrackingTouch(SeekBar seekBar) {

        }

        @Override
        public void onStopTrackingTouch(SeekBar seekBar) {

        }
    }
}

File: app/src/main/java/zone/com/zanimate/_ReboundActivity.kt
package zone.com.zanimate

import android.animation.ArgbEvaluator
import android.graphics.Color
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.SeekBar
import android.widget.TextView
import com.facebook.rebound.SimpleSpringListener
import com.facebook.rebound.Spring
import com.facebook.rebound.SpringChain
import com.facebook.rebound.SpringConfig
import com.facebook.rebound.SpringSystem
import com.facebook.rebound.SpringUtil
import com.zone.lib.utils.view.ViewUtils
import com.zone.view.SquareImageView2
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import kotlinx.android.synthetic.main.a_rebound.*

class _ReboundActivity : AppCompatActivity() {

    lateinit var spring: Spring
    private var tensionValue = SpringConfig.defaultConfig.tension.toInt()
    private var frictionValue = SpringConfig.defaultConfig.friction.toInt()


    internal var startColor = Color.argb(255, 255, 64, 230)
    internal var endColor = Color.argb(255, 0, 174, 255)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.a_rebound)
        ButterKnife.bind(this)
        spring = SpringSystem.create().createSpring()

        addReboundListener()
        updateConfig()
        updateSeekbar()

    }

    private fun addReboundListener() {
        spring.addListener(object : SimpleSpringListener() {
            override fun onSpringUpdate(spring: Spring?) {
                super.onSpringUpdate(spring)
                val currentValue = spring!!.currentValue
                val valueMap = SpringUtil.mapValueFromRangeToRange(currentValue, 0.0, 1.0, 1.0, 0.5)
                siv.scaleX = valueMap.toFloat()
                siv.scaleY = valueMap.toFloat()
            }
        })

        siv.setOnTouchListener { v, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> spring.endValue = 1.0
                MotionEvent.ACTION_UP -> spring.endValue = 0.0
            }
            true
        }
    }

    @OnClick(R.id.playTogether, R.id.play)
    fun onClick(view: View) {
        when (view.id) {
            R.id.play -> {
                siv!!.bringToFront()
                spring.endValue = (if (spring.endValue == 0.0) 1 else 0).toDouble()
            }
            R.id.playTogether -> playTogether()
        }
    }

    private fun playTogether() {
        ll_main!!.bringToFront()
        ll_main!!.removeAllViews()
        val mSpringChain = SpringChain.create()
        val offset = (ll_main!!.rootView.width + ll_main!!.width) / 2
        for (i in 0..6) {
            val v = View(this)
            val argbEvaluator = ArgbEvaluator()
            v.setBackgroundColor(argbEvaluator.evaluate(i.toFloat() / 7.toFloat(), startColor, endColor) as Int)
            ll_main!!.addView(v, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1f))
            v.translationX = (-offset).toFloat()
            mSpringChain.addSpring(object : SimpleSpringListener() {
                override fun onSpringUpdate(spring: Spring?) {
                    val value = spring!!.currentValue.toFloat()
                    v.translationX = value
                }

                override fun onSpringAtRest(spring: Spring?) {
                    super.onSpringAtRest(spring)
                    v.visibility = View.INVISIBLE
                }
            })
            if (i == 6)
                v.postDelayed({
                    val springList = mSpringChain.allSprings
                    for (spring1 in springList)
                        spring1.currentValue = (-offset).toDouble()
                    for (i1 in 0 until ll_main!!.childCount)
                        ViewUtils.recurrenceClipChildren(ll_main!!.getChildAt(i1), false)
                    mSpringChain.setControlSpringIndex(2).controlSpring.endValue = 0.0
                }, 500)
        }
    }


    private fun updateSeekbar() {
        seekTension.setOnSeekBarChangeListener(object : SimpleOnSeekBarChangeListener() {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                super.onProgressChanged(seekBar, progress, fromUser)
                tensionValue = progress
                updateConfig()
            }
        })
        seekBarFriction.setOnSeekBarChangeListener(object : SimpleOnSeekBarChangeListener() {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                super.onProgressChanged(seekBar, progress, fromUser)
                frictionValue = progress
                updateConfig()
            }
        })
    }

    private fun updateConfig() {
        tv_Tension.text = " 拉力(tension):$tensionValue"
        tv_Friction.text = " 摩擦力(friction):$frictionValue"
        seekTension.progress = tensionValue
        seekBarFriction!!.progress = frictionValue
        spring.springConfig = SpringConfig(tensionValue.toDouble(), frictionValue.toDouble())
    }


    open class SimpleOnSeekBarChangeListener : SeekBar.OnSeekBarChangeListener {
        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {

        }

        override fun onStartTrackingTouch(seekBar: SeekBar) {

        }

        override fun onStopTrackingTouch(seekBar: SeekBar) {

        }
    }
}
